### PR TITLE
Convert excerpt to Preact (1/n) - Add a utility to observe element size changes

### PR DESCRIPTION
**Pull request by robertknight**
Add a utility for use by a new implementation of the `<excerpt>`
component which observes changes in the size of the DOM output directly.

This uses `ResizeObserver` where available with a fallback to
`MutationObserver` + selected window events in other browsers.

Part of #1202

### CODE REVIEW

File: src/sidebar/util/observe-element-size.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      prevWidth = element.clientWidth;
+      prevHeight = element.clientHeight;
+      onSizeChanged(prevWidth, prevHeight);
+    }
+  };
+
+  element.addEventListener('load', check);
+  window.addEventListener('resize', check);
+  const observer = new MutationObserver(check);
+  observer.observe(element, {
```
Comment (lyzadanger) [inline src/sidebar/util/observe-element-size.js]: It can be a fine line of figuring out when to introduce comments for external APIs, etc., but I'm wondering if a quick comment here would help for future devs in terms of which `MutationObserverInit` properties have been configured here. If I am parsing this correctly, it's something like "Whenever text content changes, child nodes are added/removed or child-node text content changes"?

File: src/sidebar/util/observe-element-size.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    characterData: true,
+    childList: true,
+    subtree: true,
+  });
+
+  return () => {
+    element.removeEventListener('load', check);
+    window.removeEventListener('resize', check);
+    observer.disconnect();
```
Comment (lyzadanger) [inline src/sidebar/util/observe-element-size.js]: Good cleanup here...

File: src/sidebar/util/test/observer-element-size-test.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      await waitForObservations();
+      assert.notCalled(sizeChanged);
+    });
+  });
+
+  context('when `ResizeObserver` is not available', () => {
+    let origResizeObserver;
+    beforeEach(() => {
+      origResizeObserver = window.ResizeObserver;
+      window.ResizeObserver = undefined;
```
Comment (lyzadanger) [inline src/sidebar/util/test/observer-element-size-test.js]: Good, good good. I was wondering how this would be handled (testing fallback)...

File: src/sidebar/util/test/observer-element-size-test.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    stopObserving();
+    content.remove();
+  });
+
+  function startObserving() {
+    stopObserving = observeElementSize(content, sizeChanged);
+  }
+
+  context('when `ResizeObserver` is available', function() {
+    if (typeof ResizeObserver === 'undefined') {
```
Comment (lyzadanger) [inline src/sidebar/util/test/observer-element-size-test.js]: Can you help me understand this branch? When would we expect this condition to be truthy? When tests are being run in unexpected environments? And are there potential risks to this set of tests being skipped and/or us not noticing?

File: src/sidebar/util/test/observer-element-size-test.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      content.innerHTML = '<p>different content</p>';
+      await waitForObservations();
+      assert.called(sizeChanged);
+
+      stopObserving();
+      sizeChanged.reset();
+
+      content.innerHTML = '<p>other content</p>';
+      await waitForObservations();
+      assert.notCalled(sizeChanged);
```
Comment (lyzadanger) [inline src/sidebar/util/test/observer-element-size-test.js]: Derp: I can't seem to follow why `sizeChanged` would _not_ have been called here. Help?

File: src/sidebar/util/observe-element-size.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Watch for changes in the size (`clientWidth` and `clientHeight`) of
+ * an element.
+ *
+ * Returns a cleanup function which should be called to remove observers when
+ * updates are no longer needed.
+ *
+ * @param {Element} element - HTML element to watch
+ * @param {(width: number, height: number) => any} onSizeChanged -
+ *   Callback to invoke with the `clientWidth` and `clientHeight` of the
```
Comment (lyzadanger) [inline src/sidebar/util/observe-element-size.js]: It occurs to me to wonder why, in the current application of this for excerpt, we need `clientWidth` as well as `clientHeight`? Symmetry? Extensibility? Or actual need I'm not thinking of?

File: src/sidebar/util/observe-element-size.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      prevWidth !== element.clientWidth ||
+      prevHeight !== element.clientHeight
+    ) {
+      prevWidth = element.clientWidth;
+      prevHeight = element.clientHeight;
+      onSizeChanged(prevWidth, prevHeight);
+    }
+  };
+
+  element.addEventListener('load', check);
```
Comment (lyzadanger) [inline src/sidebar/util/observe-element-size.js]: In this fallback variant with `MutationObserver`, you're also explicitly listening for `load` and `resize` events. Are these handled more automagically by `ResizeObserver`? I note that these events are tested for explicitly only on the "fallback" branch of the tests. Do we need tests for these events on the `ResizeObserver` branch as well, to make sure they're taken care of?

File: src/sidebar/util/test/observer-element-size-test.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+      },
+      {
+        description: "the element's DOM structure changes",
+        triggerCheck: () => (content.innerHTML += '<p>more content</p>'),
+      },
+    ].forEach(({ description, triggerCheck }) => {
+      it(`checks for changes when ${description}`, async () => {
+        startObserving();
+
+        // Change the content height, which is not directly observed.
```
Comment (lyzadanger) [inline src/sidebar/util/test/observer-element-size-test.js]: "Content height" is not directly observed—because `clientHeight` is what's being observed? I admit to being a little muddled between this and `minHeight` and...anyway...can you help me understand here? Is it basically that an initial `height` of some sort needs to be set so that the observer can see a change? Sorry I'm not following!

File: src/sidebar/util/test/observer-element-size-test.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        await waitForObservations();
+        assert.called(sizeChanged);
+
+        sizeChanged.reset();
+        stopObserving();
+
+        content.style.minHeight = '200px';
+        triggerCheck();
+        await waitForObservations();
+        assert.notCalled(sizeChanged);
```
Comment (lyzadanger) [inline src/sidebar/util/test/observer-element-size-test.js]: Same comment as earlier—not following why this change would not fire the callback.

File: src/sidebar/util/observe-element-size.js (hunk 6 above)
Comment (lyzadanger) [reply src/sidebar/util/observe-element-size.js]: Actually, nevermind. The name of the module explains it 😁

File: src/sidebar/util/observe-element-size.js (hunk 7 above)
Comment (robertknight) [reply src/sidebar/util/observe-element-size.js]: `ResizeObserver` monitors when an element's size changes, regardless of cause. Listening for these DOM events is only necessary in the fallback case when we can't monitor element size changes directly but need to listen for other things which might result in size changes.

File: src/sidebar/util/observe-element-size.js (hunk 1 above)
Comment (robertknight) [reply src/sidebar/util/observe-element-size.js]: > If I am parsing this correctly, it's something like "Whenever text content changes, child nodes are added/removed or child-node text content changes"?

That is exactly right. I honestly think a comment would add noise since this is straightforward use of the API and MDN [has good documentation](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver).

File: src/sidebar/util/test/observer-element-size-test.js (hunk 4 above)
Comment (robertknight) [reply src/sidebar/util/test/observer-element-size-test.js]: By default tests run in headless Chrome where `ResizeObserver` is available. However you can manually run tests in any other browser locally by opening the Karma URL (printed when tests start running) in that browser. The only way this will fail is if some other test were to explicitly overwrite this global. In that case though the test runner would notify the user that some tests have been skipped.

File: src/sidebar/util/test/observer-element-size-test.js (hunk 5 above)
Comment (robertknight) [reply src/sidebar/util/test/observer-element-size-test.js]: The call to `stopObserving` a few lines up removes the observer.

File: src/sidebar/util/observe-element-size.js (hunk 6 above)
Comment (robertknight) [reply src/sidebar/util/observe-element-size.js]: It is true that we only care about height for the initial use case, but I made it a little more generic as that was trivial to do.

File: src/sidebar/util/test/observer-element-size-test.js (hunk 8 above)
Comment (robertknight) [reply src/sidebar/util/test/observer-element-size-test.js]: In the fallback case, the utility monitors the DOM _structure_ but cannot directly monitor the _laid out DOM elements_, which is what is changed by modifying the element's style. This is a difference from when `ResizeObserver` is available, which can monitor all causes of element size changes.

File: src/sidebar/util/test/observer-element-size-test.js (hunk 9 above)
Comment (robertknight) [reply src/sidebar/util/test/observer-element-size-test.js]: See my reply above - in the fallback case, it is known/expected that element size changes due to style changes only (ie. not DOM structure changes) are not detected.

**Comment by robertknight** [comment]
Hey @lyzadanger - I've responded to your questions above. Please let me know if that clarifies things.
